src/main/java/io/supertokens/session/Session.java

```diff
@@ -336,21 +363,29 @@ private static SessionInformationHolder refreshSessionHelper(Main main, String r
 
                         if (sessionInfo.refreshTokenHash2.equals(Utils.hashSHA256(Utils.hashSHA256(refreshToken)))) {
                             // at this point, the input refresh token is the parent one.
+
+                            JsonObject grantPayload = sessionInfo.grants;
+                            // TODO(mihaly): we could skip updating this in DB until a value is set
+                            if (grantPayload == null) {
```

Comment:
I think we should normalise grant to always at least be an empty object (and never null) when:
- fetching from db and it is null when read
- normalise the API input right at the start

We want to minimise the number of `grantlPayload == null` checks in the main logic here.

Reply:
These normalizations are only happening when we are upgrading from V2 to V3.
- API input: either it is required (In 2.13) or a null value just means no update.
- Fetching from DB: `grantPayload == null` is the way we can tell if a token is V2 or V3, so it should be conserved